qutebrowser/completion/models/histcategory.py

```diff
@@ -88,15 +88,21 @@ def set_pattern(self, pattern):
         if not self._query or len(words) != len(self._query.bound_values()):
             # if the number of words changed, we need to generate a new query
             # otherwise, we can reuse the prepared query for performance
-            self._query = sql.Query(' '.join([
-                "SELECT url, title, {}".format(timefmt),
-                "FROM CompletionHistory",
-                # the incoming pattern will have literal % and _ escaped
-                # we need to tell sql to treat '\' as an escape character
-                'WHERE ({})'.format(where_clause),
-                self._atime_expr(),
-                "ORDER BY last_atime DESC",
-            ]), forward_only=False)
+            try:
+                self._query = sql.Query(' '.join([
```

Comment:
Since you are worried about performance would having this multiline string be a class attribute and then just using `.format()` on it be faster? I'm pretty sure it would, technically at least, because it would avoid the join and and do half as many formats.

Reply:
I'll turn this down for now as I'm trying to keep the diff small to make it easier to merge (just fixing the crash), but that's one of the things I want to fix later on :wink:

Reply:
Regarding class attribute, strings only take 1 command (`LOAD_CONST`) to execute no matter how long it is, so moving it to a class attribute won't help.

Something like this would suffice, if it's clean enough:

```
self._query = sql.Query(
	"SELECT url, title, {}"
	"FROM CompletionHistory"
	# the incoming pattern will have literal % and _ escaped
	# we need to tell sql to treat '\' as an escape character
	'WHERE ({})'
	'{}'
	"ORDER BY last_atime DESC".format(timefmt, where_clause, self._atime_expr())
, forward_only=False)
```

This wont' make the diff larger anyway.

Reply:
I'm trying to minimize the logic changed, not the lines changed (to avoid introducing any bugs). For example, that example above would break this query, because it won't have spaces between lines (`'SELECT url, title, {}FROM CompletionHistoryWHERE ({}){}ORDER BY last_atime DESC'`).
That approach would work to avoid the concat on that step though :)